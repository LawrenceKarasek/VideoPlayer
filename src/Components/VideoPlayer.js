import React, { useState, useEffect, useRef, Fragment } from 'react';
import VideoControls from './VideoControls';
import PropTypes from 'prop-types';
import '../styles.css';

const VideoPlayer = ({ videos }) => {
  const [videoIndex, setVideoIndex] = useState(0);
  const [videoProgress, setVideoProgress] = useState(0);
  const [duration, setDuration] = useState(0);
  const [isPlaying, setIsPlaying] = useState(false);
  const { videoSrc, link } = videos[videoIndex];
  const videoRef = useRef();
  const intervalRef = useRef();

  let currentPercentage;
  const videoStyling = `
  -webkit-gradient(linear, 0% 0%, 100% 0%, color-stop(${currentPercentage}, #fff), color-stop(${currentPercentage}, #777))
`;

  const onDurationChangeHandler = e => {
    const seconds = Math.floor(e.target.duration);
    setDuration(seconds);
  };

  const onEnded = () => {
    clearInterval(intervalRef.current);
    toNextVideo();
  };

  const startTimer = () => {
    clearInterval(intervalRef.current);

    intervalRef.current = setInterval(() => {
      if (videoRef.current) {
        setVideoProgress(videoRef.current.currentTime);
      }
    }, [1000]);
  };

  const onScrub = value => {
    clearInterval(intervalRef.current);
    videoRef.current.currentTime = value;
    setVideoProgress(videoRef.current.currentTime);
  };

  const onScrubEnd = () => {
    if (!isPlaying) {
      setIsPlaying(true);
    }
    startTimer();
  };

  const toPrevVideo = () => {
    setIsPlaying(false);
    if (videoIndex - 1 < 0) {
      setVideoIndex(videos.length - 1);
    } else {
      setVideoIndex(videoIndex - 1);
    }
  };

  const toNextVideo = () => {
    setIsPlaying(false);
    if (videoIndex < videos.length - 1) {
      setVideoIndex(videoIndex + 1);
    } else {
      setVideoIndex(0);
    }
  };

  useEffect(() => {
    if (isPlaying) {
      videoRef.current.play();
      setVideoProgress(videoRef.current.currentTime);
      startTimer();
    } else {
      videoRef.current.pause();
    }
  }, [isPlaying]);

  useEffect(() => {
    if (!isPlaying) {
      videoRef.current.pause();
      videoRef.current.src = videoSrc;
      setIsPlaying(true);
    }
  }, [videoIndex, videoSrc]);

  return (
    <Fragment>
      {videos && (
        <div className="video-player">
          <div className="video-info">
            <label data-test-id="source-test-id">Source </label>
            <a className="link" data-test-id="link" href={link}>
              {link}
            </a>
          </div>
          <div className="video-info">
            <video
              ref={videoRef}
              className="video"
              onDurationChange={onDurationChangeHandler}
              onEnded={onEnded}
              poster={'/giphyloading.gif'}
              autoPlay={true}
              muted={true}
            ></video>
            <VideoControls
              isPlaying={isPlaying}
              onPrevClick={toPrevVideo}
              onNextClick={toNextVideo}
              onPlayPauseClick={setIsPlaying}
            />
          </div>
          {duration && (
            <input
              type="range"
              value={videoProgress}
              step="1"
              min="0"
              max={duration}
              className="progress"
              onChange={e => onScrub(e.target.value)}
              onMouseUp={onScrubEnd}
              onKeyUp={onScrubEnd}
              style={{ background: videoStyling }}
            />
          )}
        </div>
      )}
    </Fragment>
  );
};

VideoPlayer.propTypes = {
  videos: PropTypes.array,
};

export default VideoPlayer;
