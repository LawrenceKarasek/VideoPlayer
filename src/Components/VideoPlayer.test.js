import ShallowRenderer from 'react-test-renderer/shallow';
import VideoPlayer from './VideoPlayer';

it('shows video info', async () => {
  const videos = [
    {
      videoSrc: 'https://cdn.coverr.co/videos/coverr-a-dog-panting-outdoors-2248/1080p.mp4',
      link: 'https://coverr.co/videos/a-dog-panting-outdoors-C5qfQ2fp5V',
    },
    {
      videoSrc: 'https://cdn.coverr.co/videos/coverr-dog-staring-at-a-latte-5538/1080p.mp4',
      link: 'https://coverr.co/videos/dog-staring-at-a-latte-RhOkNJByHc',
    },
  ];

  const renderer = new ShallowRenderer();
  renderer.render(<VideoPlayer videos={videos} />);
  const result = renderer.getRenderOutput();

  expect(result).toMatchSnapshot();
});
