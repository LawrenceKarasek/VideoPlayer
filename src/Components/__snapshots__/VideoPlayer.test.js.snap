// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shows video info 1`] = `
<React.Fragment>
  <div
    className="video-player"
  >
    <div
      className="video-info"
    >
      <label
        data-test-id="source-test-id"
      >
        Source 
      </label>
      <a
        className="link"
        data-test-id="link"
        href="https://coverr.co/videos/a-dog-panting-outdoors-C5qfQ2fp5V"
      >
        https://coverr.co/videos/a-dog-panting-outdoors-C5qfQ2fp5V
      </a>
    </div>
    <div
      className="video-info"
    >
      <video
        autoPlay={true}
        className="video"
        muted={true}
        onDurationChange={[Function]}
        onEnded={[Function]}
        poster="/giphyloading.gif"
      />
      <VideoControls
        isPlaying={false}
        onNextClick={[Function]}
        onPlayPauseClick={[Function]}
        onPrevClick={[Function]}
      />
    </div>
    0
  </div>
</React.Fragment>
`;
